keps/sig-network/4444-service-routing-preference/README.md

```diff
@@ -0,0 +1,945 @@
+  `routingPreference` field will allow the customer to use the same Helm Chart
+  configurations with greater confidence regardless of the underlying Kubernetes
+  implementation. This simplifies their deployment process and reduces the
+  complexity of managing cross-cluster applications.
+
+#### Story 5
+* **Requirement:** I have some other precise preferences for how traffic should
+  be routed, and I know that my chosen implementation supports the desired
+  preference. 
+* **Solution:** Set `routingPreference=<domain>/<heuristicName>` (where
+  `<domain>` and `<heuristicName>` are provided by your implementation).
+* **Effect:** The Kubernetes implementation will apply the specified routing
+  heuristic. It's important to note that the precise behavior of
+  implementation-specific heuristics might vary.
+
+### Notes/Constraints/Caveats (Optional)
+
+This proposal is our third attempt at an API revolving around such a
+configuration. There's a non-zero chance that we may need to revisit this again.
+
+### Risks and Mitigations
+
+* **Risk:** Having a routing preference like `Zone` comes at the risk of
+  endpoints in certain zones being overloaded if the originating traffic is
+  skewed towards a particular zone.
+
+  **Mitigation:**
+    * Emphasize in the documentation that the `Zone` preference is
+      designed for low-latency or monetory-cost reasons, with the understanding
+      that it can lead to overload within zones.
+    * Recommend approaches like having deployments per zone which can scale
+      independently of other zones.
+
+* **Risk:** Users migrating from the `service.kubernetes.io/topology-mode`
+  annotation might encounter differences in exact routing behavior:
+  * The new field doesn't support a routing preference that is exactly similar
+    to using `service.kubernetes.io/topology-mode=Auto` from the old annotation.
+  * If both field and the annotation are set, the annotation will take
+    precedence. (However, this behavior is temporary as the annotation will be
+    deprecated and removed in future releases)
+
+  **Mitigation:** Properly document the suggested migration paths with
+  limitations.
+
+## Design Details
+
+### Standard Heuristic Implementation (kube-proxy dataplane)
+
+kube-proxy (along with EndpointSlice controller, within kube-controller-manager
+as the control plane) will support the three standard routing preferences
+(`Default`, `Spread`, `Zone`).
+
+#### `Default` and `Spread`
+* Initially, kube-proxy will treat the `Default` preference the same as
+  `Spread`
+* This leverages existing implementation, requiring no major changes.
+
+#### `Zone`
+* This preference will be implemented by the use of Hints within EndpointSlices.
+* We already use Hints to implement `service.kubernetes.io/topology-mode=Auto`
+  Similarly, we’ll use the same Hints within the EndpointSlice to implement the
+  `Zone` heuristic – the hints will match the zone of the endpoint itself.
```

Comment:
So how will `routingPreference: Zone` address the aforementioned Limitations of Topology-Aware Hints?

Reply:
I think the biggest benefit of this simpler approach is the predictability and simplicity of this approach. It will also work equally well for Services with small numbers of endpoints. If I were to summarize the feedback I've gotten about the current approach, it would largely fall into some combination of the following:

1. Why aren't hints being set?
2. How do I ensure that hints will be reliably set?

This approach should avoid both of those common points of frustration.

Reply:
> It seems like CPU-proportional is a more or less failed experiment.

Oh... I see. When the KEP says "the hints will match the zone of the endpoint itself", it's talking about _making a change_ to kcm, not explaining the current behavior of kcm. That is not clear (and is the source of my confusion below as well).